In [34]:
import pandas as pd  ## usando biblioteca pandas para explorar os dados
data = pd.read_csv("drive/MyDrive/Colab Notebooks/dados_ficha_a_desafio.csv") ## lendo o arquivo .csv via pandas
data.info() ## visualizando a dimensao e caracteristica de cada coluna

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 35 columns):
 #   Column                                 Non-Null Count   Dtype  
---  ------                                 --------------   -----  
 0   id_paciente                            100000 non-null  object 
 1   sexo                                   100000 non-null  object 
 2   obito                                  100000 non-null  object 
 3   bairro                                 100000 non-null  object 
 4   raca_cor                               100000 non-null  object 
 5   ocupacao                               100000 non-null  object 
 6   religiao                               100000 non-null  object 
 7   luz_eletrica                           100000 non-null  object 
 8   data_cadastro                          100000 non-null  object 
 9   escolaridade                           100000 non-null  object 
 10  nacionalidade                          100000 non-null  o

Como primeiro contato, podemos explorar as dimensoes do dataset: sao 35 variaveis para um total de 100.000 registros. Idealmente, teriamos todas as colunas preenchidas 100% sem objetos 'null' ou nulos, entretanto, para as variaveis 'identidade_genero', 'altura', 'peso', 'pressao_sistolica' e 'pressao_diastolica' encontramos registros em branco. Manteremos atençao a esses dados para futuras oportunidades de processamento. Podemos ver tambem que o dataset consiste em maioria de dados do tipo 'string' com poucas colunas identificadas como numeros 'int64/float64'.


In [35]:
pd.set_option('display.max_columns', None) ## impedindo que o pandas encurte as colunas quando printar
data.head() ## visualizando algumas entradas para entender os registros, impedindo o encurtamento de colunas

id_paciente    sexo  obito           bairro  \
0  cd6daa6e-038d-4952-af29-579e62e07f97    male      0  Parada de Lucas   
1  ad6cecb2-3a44-49ab-b2f3-6f9ffc2e2ec7    male      0      Cidade Nova   
2  54e834e7-e722-4daa-8909-cf917a1247e6    male      0       Santa cruz   
3  c6a71e5e-0933-48d1-9d5a-8f448dc37f71  female  False            Bangu   
4  d5262a3c-e5d3-4195-a46b-0acd2533e1d6    male      0       Santíssimo   

  raca_cor                          ocupacao      religiao luz_eletrica  \
0   Branca                     Não se aplica  Sem religião         True   
1    Parda         Assistente Administrativo    Evangélica            1   
2    Parda                     Não se aplica      Católica            1   
3   Branca  Representante Comercial Autônomo  Sem religião            1   
4   Branca               Técnico Eletricista         Outra         True   

             data_cadastro            escolaridade nacionalidade  \
0  2021-06-24 00:00:00.000  Fundamental Incompleto    Brasileira   
1  2021-01-07 00:00:00.000        Médio Incompleto    Brasileira   
2  2021-02-18 00:00:00.000    Fundamental Completo    Brasileira   
3      2022-03-09 10:40:37            Alfabetizado    Brasileira   
4      2022-12-13 18:30:45          Médio Completo    Brasileira   

       renda_familiar data_nascimento em_situacao_de_rua  frequenta_escola  \
0  2 Salários Mínimos      1996-06-25                  0                 0   
1    1 Salário Mínimo      1952-12-20                  0                 0   
2    1 Salário Mínimo      1931-11-14                  0                 0   
3  1/2 Salário Mínimo      1988-12-06                  0                 0   
4    1 Salário Mínimo      1971-01-26                  0                 1   

             meios_transporte doencas_condicoes identidade_genero  \
0                      Ônibus        ["Outros"]               Cis   
1                      Ônibus       Hipertensão               Não   
2  Carro, Metrô, Ônibus, Trem                []               Não   
3               Carro, Ônibus        ["Outros"]               Não   
4               Metrô, Ônibus                []               Não   

                                   meios_comunicacao orientacao_sexual  \
0                                Internet, Televisão     Heterossexual   
1                                Internet, Televisão     Heterossexual   
2                                                 []     Heterossexual   
3  ["R\u00e1dio", "Televis\u00e3o", "Jornal", "In...     Heterossexual   
4           ["Televis\u00e3o", "Internet", "Jornal"]     Heterossexual   

  possui_plano_saude              em_caso_doenca_procura  \
0              False                    Hospital Público   
1                  1                                  []   
2                  0  Hospital Público, Unidade de Saúde   
3                  0                    Hospital Público   
4                  0                                  []   

             situacao_profissional vulnerabilidade_social  \
0                   Emprego Formal                      0   
1  Autônomo com previdência social                      0   
2                   Emprego Formal                      0   
3                    Não se aplica                      0   
4                     Não trabalha                      0   

  data_atualizacao_cadastro familia_beneficiaria_auxilio_brasil  \
0   2022-07-08 22:45:09.710                                   0   
1   2022-07-10 17:48:09.600                                   1   
2       2023-05-06 09:19:33                                   0   
3   2022-03-15 09:35:04.550                                   1   
4       2022-07-03 22:52:09                                   0   

  crianca_matriculada_creche_pre_escola  altura  peso  pressao_sistolica  \
0                                     0   172.0  52.5              110.0   
1                                     0   158.0  76.2              140.0   
2                            

Visualizando agora as primeiras entradas do dataset, podemos inferir alguns possíveis problemas que vamos investigar para avaliar a integridade dos dados:
- As colunas 'obito', 'luz_eletrica', 'em_situacao_de_rua', 'frequenta_escola', 'possui_plano_saude', 'vulnerabilidade_social', 'familia_beneficiaria_auxilio_brasil' e 'crianca_matriculada_creche_pre_escola' possuem entradas do tipo booleano 0 ou 1. Podemos ver que algumas dessas colunas possuem misturas com entradas 'True' e 'False', sendo necessário traduzir essas entradas para booleanos correspondentes (0 = False, 1 = True). Tambem sabemos que os dados dessas colunas, com excecao de 'frequenta_escola', estao como type 'object', que poderia ser tratado para 'int64' apos a traducao;
- Colunas com 'strings' como 'doencas_condicoes', 'meios_comunicacao', 'em_caso_doenca_procura' e 'em_caso_doenca_procura' denunciam a ma formatacao de inputs para essas variaveis, indicando a necessidade de investigar e aplicar formatacao para todas essas colunas;
- Registros de 'identidade_genero' como 'Nao' precisam ser investigadas se querem realmente dizer 'Nao informado';
- As variaveis 'altura' e 'peso' mostram algumas inconsistencias com uma pessoa medindo 53 centimetros e pesando 82 quilos (provavelmente um erro na altura, podendo estar em pes e polegadas por ex.) e uma pessoa com 154cm de altura e pesando apenas 8.5 quilos (muito provavelmente erro de casa decimal, 8.5 ao inves de 85.0);

In [36]:
data.describe() ## visualizando dados numericos

frequenta_escola        altura          peso  pressao_sistolica  \
count     100000.000000  99975.000000  99816.000000       99960.000000   
mean           0.158190    142.693589     63.801629         129.544068   
std            0.364921     38.659345     35.594173          21.406429   
min            0.000000      0.000000      0.000000           0.000000   
25%            0.000000    145.000000     50.700000         120.000000   
50%            0.000000    158.000000     67.900000         130.000000   
75%            0.000000    165.000000     82.000000         140.000000   
max            1.000000    810.000000    998.000000         900.000000   

       pressao_diastolica  n_atendimentos_atencao_primaria  \
count        99983.000000                    100000.000000   
mean            79.378207                         5.542710   
std             21.062140                         5.204464   
min              0.000000                         0.000000   
25%             70.000000                         0.000000   
50%             80.000000                         5.000000   
75%             87.000000                         9.000000   
max            921.000000                        32.000000   

       n_atendimentos_hospital  
count            100000.000000  
mean                  7.312930  
std                   9.801987  
min                   0.000000  
25%                   0.000000  
50%                   2.000000  
75%                  13.000000  
max                  77.000000

Focando nos dados com entrada numérica, podemos ver que 'altura', 'peso', 'pressao_sistolica' e 'pressao_diastolica' possuem valores surreais apresentados nas categorias 'max' e 'min', como por exemplo uma pessoa com 810 centimetros de altura ou pressao sistolica de 900.


In [37]:
## vetor de colunas de interesse e frequencia de valores unicos (nao-duplicados)
cols = ['sexo', 'bairro', 'raca_cor', 'religiao', 'escolaridade', 'nacionalidade', 'renda_familiar', 'meios_transporte', 'identidade_genero', 'meios_comunicacao', 'orientacao_sexual', 'em_caso_doenca_procura', 'situacao_profissional']
data[cols].nunique()

sexo                        2
bairro                    748
raca_cor                    6
religiao                   16
escolaridade               11
nacionalidade               3
renda_familiar              9
meios_transporte          655
identidade_genero          10
meios_comunicacao         325
orientacao_sexual           4
em_caso_doenca_procura    168
situacao_profissional      13
dtype: int64

A partir da contagem das entradas unicas de algumas variaveis que ainda nao foram analisadas, podemos identificar colunas que tem um numero nao realista de respostas:
- Variaveis 'meios_comunicacao', 'meios_transporte' e 'em_caso_doenca_procura' sofrem do erro de formatacao do texto e descrevem, por exemplo, 655 formas de se locomover;
- Variaveis 'religiao', 'identidade de genero', 'escolaridade' e 'renda_familiar' podem conter mais entradas que o necessario para registro, como por exemplo duplicar resposta de '1 Salario' com '1 Salario minimo' (esse tipo de verificacao deve ser feito para todas as variaveis descritivas, mesmo que a contagem de valores unicos nao aponte erro a principio).

In [38]:
data['religiao'].value_counts() ## valores unicos e frequencia de registro de cada entrada

religiao
Sem religião                          35119
Católica                              28348
Outra                                 18714
Evangélica                            14798
Espírita                               1974
Religião de matriz africana             731
Não                                     174
Budismo                                  81
Judaísmo                                 32
Islamismo                                 8
Acomp. Cresc. e Desenv. da Criança        7
Candomblé                                 6
ESB ALMIRANTE                             5
10 EAP 01                                 1
ORQUIDEA                                  1
Sim                                       1
Name: count, dtype: int64

Explorando primeiramente a variavel religiao, podemos ver que as entradas 'Nao' poderiam ser alocadas para 'Sem religiao' e algumas entradas diferentes do esperado sao encontradas ao final com baixissima frequencia.

In [39]:
data['identidade_genero'].value_counts()

identidade_genero
Não                            69781
Cis                            21033
Mulher transexual               3792
Homem transexual                1939
Outro                           1343
Heterossexual                   1025
Sim                               50
Homossexual (gay / lésbica)       44
Travesti                          35
Bissexual                          6
Name: count, dtype: int64

Explorando a variavel 'identidade_genero' agora, podemos ver que algo similar acontece: 'Nao' e 'Sim' sao entradas indesejaveis para essa coluna, bem como confundimento de identidade de genero com sexualidade, como visto pelas respostas 'Heterossexual' por exemplo.

Explorando agora algumas interacoes entre variaveis que podemos fazer, podemos verificar se 'data_nascimento' condiz com 'situacao_profissional' ou 'ocupacao' por exemplo:

In [40]:
## filtrando as entradas com ate 10 anos de idade e explorando o cargo ocupacional
data.loc[data['data_nascimento'] > '2015-01-01', ['data_nascimento', 'situacao_profissional', 'ocupacao']]

data_nascimento     situacao_profissional  \
22         2015-04-30             Não se aplica   
28         2020-08-20            Emprego Formal   
37         2022-10-23            Emprego Formal   
40         2022-08-14          Emprego Informal   
55         2023-10-14  Pensionista / Aposentado   
...               ...                       ...   
99960      2018-05-16             Não se aplica   
99970      2017-12-18            Emprego Formal   
99976      2025-01-28          Emprego Informal   
99983      2022-05-16            Emprego Formal   
99991      2019-08-21            Emprego Formal   

                                ocupacao  
22        Catador de Material Reciclável  
28         Empregado doméstico arrumador  
37                    Vendedor Ambulante  
40        Motorista de ônibus Rodoviário  
55                         Não se aplica  
...                                  ...  
99960                 Vendedor Ambulante  
99970                      Não se aplica  
99976                      Não se aplica  
99983  Operador de Inspeção de Qualidade  
99991                  Vendedor Pracista  

[11196 rows x 3 columns]

Podemos ver que pessoas com ate 10 anos de idade estao registradas com 'Emprego Formal' ou com cargo, o que nao faz sentido e provavelmente e um confundimento entre o registro 'data_nascimento' com 'updated_at' ou 'data_atualizacao_cadastro'.


Dessa exploracao rapida da base de dados, podemos fazer as seguintes inferencias:
- As colunas com valores Booleanos precisam passar por testes para verificar que todas as entradas estao padronizadas (ou ate mesmo um tratamento rapido via python para garantir que todas as entradas sao do tipo 'int64' com valores 0 e 1);
- Algumas colunas descritivas possuem ma formatacao de texto, que inviabilisa analises mais profundas como por exemplo frequencia de meios de transporte utilizados pela populacao que e atendida nas unidades de saude;
- Variaveis 'altura', 'peso' e as medidas de pressao, por serem variaveis quantitativas, sofrem com erros de digitacao ou entao erros de conversao para a unidade de medida;
- As colunas descritivas, tambem, possuem algumas 'respostas livres' que acaba contaminando o dataset com entradas duplicadas ou entao erros de registro.


Algumas medidas que poderiam mitigar esses erros sao, respectivamente:
- As colunas booleanas poderiam ter uma restricao para impedir o registro de algum outro valor que fuja do padrao;
- Isso tambem vale para as colunas descritivas, que poderiam ser registradas sem a utilizacao de acentos para impedir a ma traducao entre sistemas operacionais;
- Medidas quantitativas poderiam ter um threshold de valores aceitaveis a depender da variavel (por exemplo maior peso que pode ser registrado como '600' kgs);
- Algumas colunas com 'resposta livre', como por exemplo 'identidade_genero' poderiam ser questionarios com opcoes fixas para diminuir o desentendimento entre o paciente e o coletador de dados.